### PR TITLE
[Pluggable Device] Adding more DEVICE_DEFAULT kernels.

### PR DESCRIPTION
**Pull request by kulinseth**
The checkin extends the device default kernels to
    - int32 operations
    - certain resource operations
    - TensorArray and TensorList operations

@penpornk

### CODE REVIEW

File: tensorflow/core/kernels/constant_op.cc [hunk 1]
```diff
@@ -358,6 +359,26 @@ void PlaceholderOp::Compute(OpKernelContext* ctx) {
...
+REGISTER_KERNEL_BUILDER(Name("OnesLike")
+                            .Device(DEVICE_DEFAULT)
+                            .TypeConstraint<int32>("T")
+                            .HostMemory("y"),
+                        OnesLikeOp<CPUDevice, int32>);
+REGISTER_KERNEL_BUILDER(Name("ZerosLike")
+                            .Device(DEVICE_DEFAULT)
+                            .TypeConstraint<int32>("T")
+                            .HostMemory("y"),
+                        ZerosLikeOp<CPUDevice, int32>);
```
Comment (penpornk) [inline tensorflow/core/kernels/constant_op.cc:380]: Nit: Could you please group them with the registrations for other devices, e.g., put "Fill" with other "Fill", "OnesLike" with "OnesLike", and "ZerosLike" with "ZerosLike"? It's easier when we skim for registrations for a specific kernel this way. Thank you!

File: tensorflow/core/kernels/resource_variable_ops.cc [hunk 2]
```diff
@@ -327,6 +327,18 @@ REGISTER_KERNEL_BUILDER(
 REGISTER_KERNEL_BUILDER(
     Name("VariableShape").Device(DEVICE_CPU).TypeConstraint<int64>("out_type"),
     VariableShapeOp<int64>);
+REGISTER_KERNEL_BUILDER(
+    Name("VariableShape").Device(DEVICE_DEFAULT)
+                         .TypeConstraint<int32>("out_type")
+                         .HostMemory("output")
+                         .HostMemory("input"),
```
Comment (penpornk) [inline tensorflow/core/kernels/resource_variable_ops.cc:334]: Nit: Can we list `input` first, just to follow the order in `REGISTER_OP`?
```suggestion
                         .HostMemory("input")
                         .HostMemory("output"),
```

File: tensorflow/core/kernels/resource_variable_ops.cc [hunk 3]
```diff
@@ -327,6 +327,18 @@ REGISTER_KERNEL_BUILDER(
...
+                         VariableShapeOp<int32>);
+REGISTER_KERNEL_BUILDER(
+    Name("VariableShape").Device(DEVICE_DEFAULT)
+                         .TypeConstraint<int64>("out_type")
+                         .HostMemory("output")
+                         .HostMemory("input"),
```
Comment (penpornk) [inline tensorflow/core/kernels/resource_variable_ops.cc:340]: Nit: Same here. Can we list `input` before `output`?
```suggestion
                         .HostMemory("input")
                         .HostMemory("output"),
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! I have a few nits.

cc: @jzhoulon @reedwm

File: tensorflow/core/kernels/constant_op.cc (hunk 1 above)
Comment (kulinseth) [reply tensorflow/core/kernels/constant_op.cc]: Thanks for the review @penpornk !. Done.

File: tensorflow/core/kernels/resource_variable_ops.cc (hunk 2 above)
Comment (kulinseth) [reply tensorflow/core/kernels/resource_variable_ops.cc]: Done.

File: tensorflow/core/kernels/resource_variable_ops.cc (hunk 3 above)
Comment (kulinseth) [reply tensorflow/core/kernels/resource_variable_ops.cc]: Done.

**Review by penpornk — APPROVED** [review]
Thank you again for the PR!

**Comment by kulinseth** [comment]
Hi @penpornk, @reedwm  , can you please help with the rocm and macos build issues. I looked at the logs and couldn't figure out if its related to this change.

**Pull request merged**